# Здравствуйте

Продолжаю улучшать свои навыки и знания.
Предлагаю ознакомиться с мини проектом "Анализом лояльности пользователей", в котором привожу решение простых типовых задач.

# Задачи

* Необходимо получить данные для тестирования программы лояности.
* На платформе размещены товары нескольких крупных брендов, неоходимо выбрать самых лояных клиентов, которые в долгосрочной перспективе компенсируют скидку своей лояльностью.


In [34]:
import pandas as pd

In [35]:
df = pd.read_csv('https://stepik.org/media/attachments/course/111949/lesson_3_data_1_.csv', encoding='windows-1251')

In [36]:
df.head()

Unnamed: 0                  tk  pl         dia        hs         tc  \
0        1242  120071109002055793   1  2007/11/09  0.505729  110000761   
1        1243  120070601004053633   1  2007/06/01  0.588519  110000761   
2        1244  120070928004076410   1  2007/09/28  0.785498   28491841   
3        1245  120070616003074261   1  2007/06/16  0.550961   95931501   
4        1246  120070519001018477   1  2007/05/19  0.804954   93265591   

        cta  id_art  id_subsubfam   id_subfam  ...   id_famn  id_seccion  \
0  11000076   21895  101070640100  1010706401  ...  10107064       10107   
1  11000076   21816  101070640100  1010706401  ...  10107064       10107   
2   2849184  562840  101070640100  1010706401  ...  10107064       10107   
3   9593150   28914  101070640100  1010706401  ...  10107064       10107   
4   9326559  159867  101070640100  1010706401  ...  10107064       10107   

   id_subagr  id_agr   vta  uni  id_artn                              art_sp  \
0        101       1  0.68  1.0    21895  MARAVILLA        500 G Store_Brand   
1        101       1  0.38  1.0    21816  FIDEO CABELLIN   500 G Store_Brand   
2        101       1  1.55  2.0   562840        SPAGUETTI Nє 5 500 G Brand_1   
3        101       1  1.03  2.0    28914       FIDEO FIDEUБ      500 Brand_7   
4        101       1  1.09  1.0   159867      MACARRONES  GRATINAR 5 Brand_2   

                 fam_sp fam_en  
0  PASTA ALIMENTICIA SE  pasta  
1  PASTA ALIMENTICIA SE  pasta  
2  PASTA ALIMENTICIA SE  pasta  
3  PASTA ALIMENTICIA SE  pasta  
4  PASTA ALIMENTICIA SE  pasta  

[5 rows x 21 columns]

Для удобства создам переменную user_df , поместив в нее строки tc и art_sp

In [37]:
user_df = df[['tc', 'art_sp']]

In [38]:
user_df

tc                              art_sp
0       110000761  MARAVILLA        500 G Store_Brand
1       110000761  FIDEO CABELLIN   500 G Store_Brand
2        28491841        SPAGUETTI Nє 5 500 G Brand_1
3        95931501       FIDEO FIDEUБ      500 Brand_7
4        93265591      MACARRONES  GRATINAR 5 Brand_2
...           ...                                 ...
48124    45518841      FIDEOS 0         500 G Brand_4
48125   110824211      PLUMAS 3         500 G Brand_4
48126  1408670389  MACARRONES       500 G Store_Brand
48127  1408670389  SPAGHETTI        500 G Store_Brand
48128    48582221  SPAGHETTINI      500 G Store_Brand

[48129 rows x 2 columns]

In [39]:
user_df = user_df.rename(columns={'tc': 'user_id', 'art_sp': 'brand_info'})
# переименовал в более удобные для работы названия колонок

In [40]:
user_df.head()

user_id                          brand_info
0  110000761  MARAVILLA        500 G Store_Brand
1  110000761  FIDEO CABELLIN   500 G Store_Brand
2   28491841        SPAGUETTI Nє 5 500 G Brand_1
3   95931501       FIDEO FIDEUБ      500 Brand_7
4   93265591      MACARRONES  GRATINAR 5 Brand_2

* Разделение строк - split - метод

для дальнейшей работы с данными (разделю колонку brand_info на колонки brand_name и brand_info)

In [41]:
def split_brand(brand_name_data):
    return brand_name_data.split(' ')[-1]

In [42]:
brnad_name = 'MARAVILLA 500 G Store_Brand'

In [43]:
brnad_name.split(' ')[-1]

'Store_Brand'

In [44]:
user_df['brand_name'] = user_df.brand_info.apply(split_brand)

In [45]:
# user_df['brand_name'] = user_df.brand_info.apply(lambda x: x.split(' ')[-1])
# можно использовать lambda функцию, чтобы разрезать строку по пробелу и взять последний элемент

In [46]:
user_df.head(7)

user_id                          brand_info   brand_name
0  110000761  MARAVILLA        500 G Store_Brand  Store_Brand
1  110000761  FIDEO CABELLIN   500 G Store_Brand  Store_Brand
2   28491841        SPAGUETTI Nє 5 500 G Brand_1      Brand_1
3   95931501       FIDEO FIDEUБ      500 Brand_7      Brand_7
4   93265591      MACARRONES  GRATINAR 5 Brand_2      Brand_2
5   93265591      SPAGUETI  PLATOS 500 G Brand_2      Brand_2
6  110002291  SPAGHETTINI      500 G Store_Brand  Store_Brand

**Проверю гипотезу о наличии в наших данных лояльных клиентов.**


Мне интересно узнать id_клиентов которые совершают покупки одного бренда. 
* сгрупирую клиентов по количеству покупок
* задам релевантное имя колонке "покупки" = "purchases"
* выберу клиентов которые совершили 5 или более покупок


In [47]:
users_purchases = user_df.groupby('user_id', as_index=False)\
    .agg({'brand_name': 'count'})\
    .rename(columns={'brand_name': 'purchases'})\
    .query('purchases >= 5')
users_purchases

user_id  purchases
1            1031          6
2            4241          5
11          25971          7
14          40911         27
16          45181          5
...           ...        ...
11740  1408767189          5
11750  1408783189         10
11751  1408783379          6
11754  1408798879          8
11757  1408810219          9

[3383 rows x 2 columns]

In [48]:
users_purchases.purchases.describe()
#.query('purchases') -  50% клиентов совершили 5 покупокупок
#.query('purchases >= 5') довил условие вывода клиентов совершивших 5 и более покупок
#среднеквадратическое количество покупок одного бренда = 5,32
#среднее арифметическое количество покупок = 9,32
#на мой взгляд, все логично и критерий для выборки можно оставить прежним 

count    3383.000000
mean        9.320130
std         5.623993
min         5.000000
25%         6.000000
50%         7.000000
75%        11.000000
max        60.000000
Name: purchases, dtype: float64

**Разведочный анализ данных выполнил. Перехожу к решению задачи.**
* узнаю количество покупок одного бренда для каждого из клиентов. сохраню в переменную "brand_purchases"
* Узнаю любимый бренд каждого клиента. сохраню в переменную "purchases_lovely_brand_df"

In [49]:
brand_purchases = user_df.groupby(['user_id', 'brand_name'], as_index=False) \
    .agg({'brand_info' : 'count'}) \
   
# группирую данные user_id и brand_name и считаю количество покупок каждого бренда
# получил количество покупокупок совершенных клиентов в кадом из брендов более 5 позиций
brand_purchases

user_id   brand_name  brand_info
0     -1236394515      Brand_4           1
1            1031      Brand_3           1
2            1031  Store_Brand           5
3            4241      Brand_4           3
4            4241  Store_Brand           2
...           ...          ...         ...
18183  1408817589  Store_Brand           2
18184  1408825059      Brand_1           1
18185  1408832719      Brand_4           3
18186  1408840919  Store_Brand           1
18187  1408849249  Store_Brand           1

[18188 rows x 3 columns]

In [50]:
purchases_lovely_brand_df = user_df.groupby(['user_id', 'brand_name'], as_index=False) \
    .agg({'brand_info' : 'count'}) \
    .sort_values(['user_id', 'brand_info'], ascending=[False, False])\
    .groupby('user_id') \
    .head(1)\
    .rename(columns={'brand_name' :'lovely_brand', 'brand_info' : 'lovely_brand_purchases' })

* группировал(сортировал)) клиентов и количество покупок по убыванию
* далее для каждого клиента сгруппирую максимальное количество покупок товаров одного бренда (ascending=[False, False])
* метод head() после группировки выведет указанную строку
* присвоил колонкам релевантные имена

Выведу для каждого клиента число уникальных брендов, результат сохраню в переменую users_unique_brands

In [51]:
users_unique_brands = user_df.groupby('user_id', as_index=False)\
    .agg({'brand_name': pd.Series.nunique})\
    .rename(columns={'brand_name' : 'unique_brands'})


#pd.Series.nunique -расчет количества уникальных брендов

* merge метод для сведения датафреймов
* ??pd.DataFrame.merge - вывод документации

In [52]:
users_purchases
# узнал сколько всего покупок совершают клиенты и выбрал совершивших 5 и более покупок

user_id  purchases
1            1031          6
2            4241          5
11          25971          7
14          40911         27
16          45181          5
...           ...        ...
11740  1408767189          5
11750  1408783189         10
11751  1408783379          6
11754  1408798879          8
11757  1408810219          9

[3383 rows x 2 columns]

In [53]:
users_unique_brands
# посчитал число уникальных брендов для каждого клиента

user_id  unique_brands
0     -1236394515              1
1            1031              2
2            4241              2
3           17311              1
4           17312              2
...           ...            ...
11759  1408817589              2
11760  1408825059              1
11761  1408832719              1
11762  1408840919              1
11763  1408849249              1

[11764 rows x 2 columns]

In [54]:
purchases_lovely_brand_df
# посчитал для каждого клиента число покупок товаров любимого бренда

user_id lovely_brand  lovely_brand_purchases
18187  1408849249  Store_Brand                       1
18186  1408840919  Store_Brand                       1
18185  1408832719      Brand_4                       3
18184  1408825059      Brand_1                       1
18183  1408817589  Store_Brand                       2
...           ...          ...                     ...
6           17312      Brand_1                       1
5           17311      Brand_4                       2
3            4241      Brand_4                       3
2            1031  Store_Brand                       5
0     -1236394515      Brand_4                       1

[11764 rows x 3 columns]

In [55]:
loyalty_users=users_purchases\
    .merge(users_unique_brands, on='user_id')\
    .merge(purchases_lovely_brand_df, on='user_id')

# user_id - id_покупателя
# purchases - число покупупок 
# unique_brands - число уникальных брендов в которых совершались покупки
# lovely_brand - любимый бренд
# lovely_brand_purchases - число покупок на любимый бренд
loyalty_users

user_id  purchases  unique_brands lovely_brand  \
0           1031          6              2  Store_Brand   
1           4241          5              2      Brand_4   
2          25971          7              2  Store_Brand   
3          40911         27              5      Brand_4   
4          45181          5              4  Store_Brand   
...          ...        ...            ...          ...   
3378  1408767189          5              1      Brand_4   
3379  1408783189         10              2  Store_Brand   
3380  1408783379          6              2      Brand_1   
3381  1408798879          8              3  Store_Brand   
3382  1408810219          9              2      Brand_2   

      lovely_brand_purchases  
0                          5  
1                          3  
2                          5  
3                         19  
4                          2  
...                      ...  
3378                       5  
3379                       8  
3380                       4  
3381                       4  
3382                       6  

[3383 rows x 5 columns]

**Переименную имена колонок на русский, для дальнейшего сохранения в Эксель **

In [56]:
loyalty_users_ru = loyalty_users.rename(columns={'user_id': 'id_покупателя', \
                                                 'purchases': 'число покупупок', \
                                                 'unique_brands' : 'число уникальных брендов в которых совершались покупки',\
                                                 'lovely_brand' : 'любимый бренд',\
                                                 'lovely_brand_purchases' : 'число покупок на любимый бренд'})

In [57]:
loyalty_users_top = loyalty_users[loyalty_users.unique_brands == 1]\

loyalty_users_top_ru = loyalty_users_top.rename(columns={'user_id': 'id_покупателя', \
                                                 'purchases': 'число покупупок', \
                                                 'unique_brands' : 'число уникальных брендов в которых совершались покупки',\
                                                 'lovely_brand' : 'любимый бренд',\
                                                 'lovely_brand_purchases' : 'число покупок на любимый бренд'})

# показал пользователей совершаюших покупки в одном бренде
# переименовал имена колонок на русский, для дальнейшего сохранения в эксель на лист  loyalty_users_top

**Сохраню полученные данные**
* loyalty_users_top_ru(клиенты совершившие более 5 покупок)
* loyalty_users_ru(клиенты совершившие более 5 покупки в одном бренде)

**в документ Эксель на двух листах**

In [58]:
writer = pd.ExcelWriter('loyalty_users.xlsx', engine='xlsxwriter')
loyalty_users_ru.to_excel(writer, sheet_name='loyalty_users', index=False) 
loyalty_users_top_ru.to_excel(writer, sheet_name='loyalty_users_top', index=False)
writer.save()

# К метрикам

**Интересно узнать коэффициент лояльности покупателя к бренду, для этого разделю количество покупок в одном бренде на общее количество совершенных покупок **

In [59]:
loyalty_users_top['loyalty_score'] = loyalty_users_top.lovely_brand_purchases / loyalty_users_top.purchases 

C:\Users\PROFES~1\AppData\Local\Temp/ipykernel_8932/2020356825.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  loyalty_users_top['loyalty_score'] = loyalty_users_top.lovely_brand_purchases / loyalty_users_top.purchases


In [60]:
loyalty_users_top

user_id  purchases  unique_brands lovely_brand  \
13         86281         14              1      Brand_4   
18         94961          6              1      Brand_4   
29        132061          9              1      Brand_4   
30        134281          6              1      Brand_4   
35        157311         12              1      Brand_4   
...          ...        ...            ...          ...   
3372  1010244089          9              1  Store_Brand   
3374  1010247239          5              1      Brand_4   
3376  1010274559          5              1      Brand_4   
3377  1110091379          5              1      Brand_1   
3378  1408767189          5              1      Brand_4   

      lovely_brand_purchases  loyalty_score  
13                        14            1.0  
18                         6            1.0  
29                         9            1.0  
30                         6            1.0  
35                        12            1.0  
...                      ...            ...  
3372                       9            1.0  
3374                       5            1.0  
3376                       5            1.0  
3377                       5            1.0  
3378                       5            1.0  

[911 rows x 6 columns]

**Почитал loyalty_score(коэффициент лояльности), можно продолжить анализ и посчитать медианы и процентили или построить графики и наглядно посмотреть как пользователи отличаются по лояльности**

* так как данные из выборки в результате расчетов показали 911 пользователей с коэффициентом лояльность 1
* перейду к написанию аналитической записки

# Аналитическая записка

### Итоги аналитического исследования пользователей с целью выявления максимально лояльных покупателей

* В результате подведения аналитического исследования данных 48129 пользователей, совершавших покупки в 7 брендах, мною было обнаружено 3383 пользователя совершивших 5 и более покупок, 911 пользователей совершивших 5 и более покупок товаров одного бренда.
* Для исследования применял следующие правила: процентили, среднего арифметического значения элементов массива, правила стандартного отклонения элементов в заданном массиве, а также основные математические функции и методы.
* Мной был получен список максимально лояльных пользователей с коэффициентом лояльности 1, список находится в документе Microsoft Excel (loyalty_users), на листе второй книги(loyalty_users_top), на мой взгляд данные пользователи в долгосрочной перспективе компенсируют скидку своей лояльностью.

                                                                                                            С уважением,
                                                                                                            Осокин Максим
                                                                                                            Владимирович